### PR TITLE
bugfix: create launchplan from reference workflow

### PR DESCRIPTION
**Pull request by cosmicBboy**
Signed-off-by: wild-endeavor <wild-endeavor@users.noreply.github.com>

# TL;DR
This diff aligns `Workflow` and `ReferenceWorkflow` properties that have to do with `interface`

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
 https://github.com/flyteorg/flyte/issues/802

### CODE REVIEW

File: tests/flytekit/unit/core/test_references.py [hunk 1]
```diff
@@ -292,3 +293,11 @@ def inner_test(ref_mock):
     sdk_wf = get_serializable(OrderedDict(), serialization_settings, wf1)
     assert sdk_wf.nodes[1].workflow_node.launchplan_ref.project == "proj"
     assert sdk_wf.nodes[1].workflow_node.launchplan_ref.name == "app.other.flyte_entity"
+
+
+def test_lp_from_ref_wf():
+    @reference_workflow(project="project", domain="domain", name="name", version="version")
+    def ref_wf1(p1: str, p2: str) -> None:
+        ...
```
Comment (cosmicBboy) [inline tests/flytekit/unit/core/test_references.py:301]: should this codecov warning be ignored?

**Comment by kumare3** [comment]
@cosmicBboy can you add #802 to the tracking issue field?

File: flytekit/core/reference_entity.py [hunk 2]
```diff
@@ -75,23 +75,22 @@ def __init__(
...
-        self._typed_interface = transform_interface_to_typed_interface(self._interface)
+        self.reset_interface(inputs, outputs)
 
     def execute(self, **kwargs) -> Any:
         raise Exception("Remote reference entities cannot be run locally. You must mock this out.")
 
-    def reset_interface(self, inputs: Dict[str, Type], outputs: Dict[str, Type]):
-        self._interface = Interface(inputs=inputs, outputs=outputs)
-        self._typed_interface = transform_interface_to_typed_interface(self._interface)
+    def reset_interface(self, inputs: Optional[Dict[str, Type]], outputs: Dict[str, Type]):
```
Comment (kumare3) [inline flytekit/core/reference_entity.py:83]: @wild-endeavor why do we have a mutation on this API? I actually do not like mutations on such an api?

File: flytekit/core/reference_entity.py [hunk 3]
```diff
@@ -114,7 +113,7 @@ def unwrap_literal_map_and_execute(
 
         # Invoked before the task is executed
         # Translate the input literals to Python native
-        native_inputs = TypeEngine.literal_map_to_kwargs(ctx, input_literal_map, self.interface.inputs)
+        native_inputs = TypeEngine.literal_map_to_kwargs(ctx, input_literal_map, self._native_interface.inputs)
```
Comment (kumare3) [inline flytekit/core/reference_entity.py:116]: Another note for @wild-endeavor, this method looks repeated from the task/workflow - call method? Do you think its better to factor it out?

File: tests/flytekit/unit/core/test_references.py (hunk 1 above)
Comment (kumare3) [reply tests/flytekit/unit/core/test_references.py:301]: I guess so, but its weird, what happens if you use `pass`

**Comment by kumare3** [comment]
lgtm, but I will let @wild-endeavor +1

File: flytekit/core/reference_entity.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/core/reference_entity.py:83]: I think I forgot to delete this.  Shouldn't be used.  @cosmicBboy can you delete?

File: flytekit/core/reference_entity.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/core/reference_entity.py:116]: Yeah but maybe not as part of this pr.  All of reference entity is copy/paste from other components - we need to clean up things generally.

File: tests/flytekit/unit/core/test_references.py [hunk 4]
```diff
@@ -292,3 +293,11 @@ def inner_test(ref_mock):
...
+
+    LaunchPlan.create("reference-wf-12345", ref_wf1, fixed_inputs={"p1": "p1-value", "p2": "p2-value"})
```
Comment (wild-endeavor) [inline tests/flytekit/unit/core/test_references.py:303]: maybe add an assert here checking the workflow id?

File: flytekit/core/reference_entity.py [hunk 5]
```diff
@@ -75,23 +75,22 @@ def __init__(
...
-    def reset_interface(self, inputs: Dict[str, Type], outputs: Dict[str, Type]):
-        self._interface = Interface(inputs=inputs, outputs=outputs)
-        self._typed_interface = transform_interface_to_typed_interface(self._interface)
+    def reset_interface(self, inputs: Optional[Dict[str, Type]], outputs: Dict[str, Type]):
+        self._native_interface = Interface(inputs=inputs, outputs=outputs)
+        self._interface = transform_interface_to_typed_interface(self._native_interface)
 
     @property
-    def interface(self) -> Interface:
+    def interface(self) -> Optional[_interface_models.TypedInterface]:
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply flytekit/core/reference_entity.py:88]: this change broke a bunch of tests in `test_references.py` that tested reference tasks, so a lot of the changes from `self.interface -> self._native_interface` were to fix those breakages. Not sure if that's what you mean by aligning?

File: tests/flytekit/unit/core/test_references.py (hunk 4 above)
Comment (cosmicBboy) [reply tests/flytekit/unit/core/test_references.py]: added tests

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**